src/reflect/scala/reflect/internal/Names.scala

```diff
@@ -14,607 +14,291 @@ package scala
-    /** The next name in the same hash bucket. */
-    def next: Name with ThisNameType
-
-    /** The length of this name. */
-    final def length: Int = len
-    final def isEmpty = length == 0
-    final def nonEmpty = !isEmpty
-
-    def nameKind: String
-    def isTermName: Boolean
-    def isTypeName: Boolean
-    def toTermName: TermName
-    def toTypeName: TypeName
-    def companionName: Name
-
-    /** Return the subname with characters from from to to-1. */
-    def subName(from: Int, to: Int): Name with ThisNameType
-    override def subSequence(from: Int, to: Int): CharSequence = subName(from, to)
-
-    /** Return a new name of the same variety. */
-    def newName(str: String): Name with ThisNameType
-
-    /** Return a new name based on string transformation. */
-    def mapName(f: String => String): Name with ThisNameType = newName(f(toString))
-
-    /** Copy bytes of this name to buffer cs, starting at position `offset`. */
-    final def copyChars(cs: Array[Char], offset: Int) =
-      java.lang.System.arraycopy(chrs, index, cs, offset, len)
-
-    /** @return the ascii representation of this name */
-    final def toChars: Array[Char] = {  // used by ide
-      val cs = new Array[Char](len)
-      copyChars(cs, 0)
-      cs
+  //TODO consider a better structure to use than a CHM
+  final def newName(value:String) = {
+    var res = cache.get(value)
+    if (res eq null) {
+      val next = new NameHolder(value)
+      res = cache.putIfAbsent(value, next)
+      if (res eq null)
+        res = next
     }
-
-    /** @return the hash value of this name */
-    final override def hashCode(): Int = index
-
-    /** @return true if the string value of this name is equal
-     *  to the string value of the given name or String.
-     */
-    def string_==(that: Name): Boolean   = (that ne null) && (toString == that.toString)
-    def string_==(that: String): Boolean = (that ne null) && (toString == that)
-
-    /****
-     *  This has been quite useful to find places where people are comparing
-     *  a TermName and a TypeName, or a Name and a String.
-
-    override def equals(other: Any) = paranoidEquals(other)
-    private def paranoidEquals(other: Any): Boolean = {
-      val cmp = this eq other.asInstanceOf[AnyRef]
-      if (cmp || !nameDebug)
-        return cmp
-
-      other match {
-        case x: String  =>
-          Console.println(s"Compared $debugString and String '$x'")
-        case x: Name    =>
-          if (this.isTermName != x.isTermName) {
-            val panic = this.toTermName == x.toTermName
-            Console.println("Compared '%s' and '%s', one term, one type.%s".format(this, x,
-              if (panic) "  And they contain the same name string!"
-              else ""
-            ))
-          }
-        case _ =>
+    res
+    //same as cache.computeIfAbsent(value, new NameHolder(_)) but faster
+  }
+  private[this] final val cache = new ConcurrentHashMap[String, NameHolder](100000, 0.75F, 1)
```

Comment:
Doubtfully your memory problem but this seems like a very big table to start out with...